From 649f42b066b0370793eba5a659da1ef04bf86478 Mon Sep 17 00:00:00 2001
From: Andrey Dolnikov <andrey.dolnikov@cogentembedded.com>
Date: Thu, 22 Oct 2020 16:15:13 +0300
Subject: [PATCH 46/76] drm/rcar_du: rework error handling in to avoid warnings

This helps to avoid the following kernel warning on V3x boards:

  WARNING: CPU: 0 PID: 5 at drivers/gpu/drm/rcar-du/rcar_du_crtc.c:878
  rcar_du_crtc_atomic_duplicate_state+0x4c/0x60

Signed-off-by: Valentine Barshak <valentine.barshak@cogentembedded.com>
---
 drivers/gpu/drm/rcar-du/rcar_du_drv.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/rcar-du/rcar_du_drv.c b/drivers/gpu/drm/rcar-du/rcar_du_drv.c
index 3503fe5..7389bbb 100644
--- a/drivers/gpu/drm/rcar-du/rcar_du_drv.c
+++ b/drivers/gpu/drm/rcar-du/rcar_du_drv.c
@@ -695,7 +695,7 @@ static int rcar_du_probe(struct platform_device *pdev)
 		if (ret != -EPROBE_DEFER)
 			dev_err(&pdev->dev,
 				"failed to initialize DRM/KMS (%d)\n", ret);
-		goto error;
+		goto modeset_err;
 	}
 
 	ddev->irq_enabled = 1;
@@ -717,7 +717,12 @@ static int rcar_du_probe(struct platform_device *pdev)
 	return 0;
 
 error:
-	rcar_du_remove(pdev);
+	drm_kms_helper_poll_fini(ddev);
+	drm_atomic_helper_shutdown(ddev);
+
+modeset_err:
+	drm_mode_config_cleanup(ddev);
+	drm_dev_put(ddev);
 
 	return ret;
 }
-- 
2.7.4

